README.md
Node CartoDB PSQL wrapper
=========================
[![Build Status](http://travis-ci.org/CartoDB/node-cartodb-psql.png)](http://travis-ci.org/CartoDB/node-cartodb-psql)

A simple wrapper for [node.js' PostgreSQL client](https://github.com/brianc/node-postgres) with specific assumptions for
CartoDB's [Windshaft](https://github.com/CartoDB/Windshaft) and [CartoDB-SQL-API](https://github.com/CartoDB/CartoDB-SQL-API)
projects.

# Requirements
 * node >= 10
 * npm
 * postgres

# Install dependencies
```sh
$ npm install
```

# Tests
```sh
$ npm test
```

## Contributing

See [CONTRIBUTING.md](CONTRIBUTING.md).


NEWS.md
## Version 0.14.0 (2019-09-10)
 - Add optional timeout parameter to `PGSQL.query`. 

## Version 0.13.1 (2018-12-10)
 - Include host and port in the dbkey, so that type cache is
   recomputed upon migration #39

## Version 0.13.0 (2018-11-20)
 - Add support for Node.js 10
 - Add package-lock.json
 - `.end()` accepts callback

## Version 0.12.0 (2018-06-06)
 - Add `psql.typeModInfo(typmod)`

## Version 0.11.0 (2018-05-28)
 - Use node-postgres to `6.4.2-cdb1`.

## Version 0.10.2 (2017-10-03)
 - Upgrade debug to 3.x.

## Version 0.10.1 (2017-08-13)
 - Upgrade node-postgres to 6.1.6-cdb1.

## Version 0.10.0 (2017-08-10)
 - Remove check for SET statements.

## Version 0.9.0 (2017-08-09)
 - Expose `pg.end` #26.

## Version 0.8.0 (2017-05-05)
 - Remove `destroyOnError` option as we need to always destroy on error #22.
 - Remove `window_sql` function.

## Version 0.7.1 (2017-01-16)
 - Remove maxRowSize after query ends or errors #15.

## Version 0.7.0 (2016-12-21)
 - Upgrades pg to 6.1.2-cdb1.
 - Upgrade dev dependencies.
 - Replace comments with a single space #9.

## Version 0.6.1 (2015-09-16)
 - Use debug module to not output to stdout by default

## Version 0.6.0 (2015-08-04)
 - Updates types from pg_type.h (#1)
 - New `destroyOnError` option to remove clients from pool after a query fails
 - Discourages node 0.8.x

## Version 0.5.1 (2015-02-20)
 - Upgrades pg to 2.6.2-cdb3: keep alive set on socket connect

## Version 0.5.0 (2015-02-19)
 - Allow to specify keep-alive config via global.settings.db_keep_alive (#3)
   - It requires to use config as object
 - Connection config can be an object or a connectionString
 - Propagate db_max_row_size from global settings to pg default's maxRowSize
   Allows to set the max length in bytes for a row

## Version 0.4.0 (2014-08-14)
 - Have PGSQL.query take an optional third argument to request read-only
   See https://github.com/CartoDB/Windshaft/issues/130
 - Send 500 status and better error message on db connection error
 - Allow to use PSQL class with default libpq parameters

## Version 0.3.1 (2014-08-11)
 - Changes pg version to emit end event on close connection event

## Version 0.3.0 (2014-08-11)
 - Adds support for cancel a query, extracts functionality from CartoDB-SQL-API

## Version 0.2.0 (2014-08-06)
 - Changes to pick pool configuration from global settings if they exist

## Version 0.1.0 (2014-05-06)
 - Initial release
 - Porting from [CartoDB-SQL-API](https://github.com/CartoDB/CartoDB-SQL-API)


run_tests.sh
#!/bin/sh

# To make output dates deterministic
export TZ='Europe/Rome'


# This is where postgresql connection parameters are read from
TESTENV=config/environments/test.js

# Extract postgres configuration
PGHOST=`node -e "console.log(require('${TESTENV}').db_host || '')"`
echo "PGHOST: $PGHOST"
PGPORT=`node -e "console.log(require('${TESTENV}').db_port || '')"`
echo "PGPORT: $PGPORT"

TESTUSERID=1

TEST_DB=`node -e "console.log(require('${TESTENV}').db_base_name || '')"`
if test -z "$TEST_DB"; then
  echo "Missing db_base_name from ${TESTENV}" >&2
  exit 1
fi
TEST_DB=`echo ${TEST_DB} | sed "s/<%= user_id %>/${TESTUSERID}/"`

cleanup() {
    dropdb ${TEST_DB} 2> /dev/null # error expected if doesn't exist
}
trap 'cleanup' 0 1 2 3 6 9 15

die() {
	msg=$1
	echo "${msg}" >&2
	cleanup
	exit 1
}

export PGHOST PGPORT

cleanup
createdb -EUTF8 ${TEST_DB} || die "Could not create test database"

# public user role
PUBLICUSER=`node -e "console.log(require('${TESTENV}').db_pubuser || 'xxx')"`
PUBLICPASS=`node -e "console.log(require('${TESTENV}').db_pubuser_pass || 'xxx')"`
echo "DROP USER IF EXISTS ${PUBLICUSER};" | psql -v ON_ERROR_STOP=1 ${TEST_DB} || exit 1
echo "CREATE USER ${PUBLICUSER} WITH PASSWORD '${PUBLICPASS}';" | psql -v ON_ERROR_STOP=1 ${TEST_DB} || exit 1

# db owner role
TESTUSER=`node -e "console.log(require('${TESTENV}').db_user || '')"`
if test -z "$TESTUSER"; then
  echo "Missing db_user from ${TESTENV}" >&2
  exit 1
fi
TESTUSER=`echo ${TESTUSER} | sed "s/<%= user_id %>/${TESTUSERID}/"`
TESTPASS=`node -e "console.log(require('${TESTENV}').db_user_pass || '')"`
TESTPASS=`echo ${TESTPASS} | sed "s/<%= user_id %>/${TESTUSERID}/"`
echo "DROP USER IF EXISTS ${TESTUSER};" | psql -v ON_ERROR_STOP=1 ${TEST_DB} || exit 1
echo "CREATE USER ${TESTUSER} WITH PASSWORD '${TESTPASS}';" | psql -v ON_ERROR_STOP=1 ${TEST_DB} || exit 1

mocha -t 5000 -u tdd $@
exit $?


test/setup.js
var _   = require('underscore'),
    env = require('../config/environments/test');

global.settings = {};
_.extend(global.settings, env);


test/run_tests.sh
#!/bin/sh

# To make output dates deterministic
export TZ='Europe/Rome'

OPT_COVERAGE=no      # run tests with coverage

while [ -n "$1" ]; do
        if test "$1" = "--with-coverage"; then
                OPT_COVERAGE=yes
                shift
                continue
        else
                break
        fi
done

cd $(dirname $0)
BASEDIR=$(pwd)
cd -

# This is where postgresql connection parameters are read from
TESTENV="${BASEDIR}/../config/environments/test.js"

# Extract postgres configuration
PGHOST=`node -e "console.log(require('${TESTENV}').db_host || '')"`
echo "PGHOST: $PGHOST"
PGPORT=`node -e "console.log(require('${TESTENV}').db_port || '')"`
echo "PGPORT: $PGPORT"

TESTUSERID=1

TEST_DB=`node -e "console.log(require('${TESTENV}').db_base_name || '')"`
if test -z "$TEST_DB"; then
  echo "Missing db_base_name from ${TESTENV}" >&2
  exit 1
fi
TEST_DB=`echo ${TEST_DB} | sed "s/<%= user_id %>/${TESTUSERID}/"`

cleanup() {
    dropdb ${TEST_DB} 2> /dev/null # error expected if doesn't exist
}
trap 'cleanup' 0 1 2 3 6 9 15

die() {
	msg=$1
	echo "${msg}" >&2
	cleanup
	exit 1
}

export PGHOST PGPORT

cleanup
createdb -EUTF8 ${TEST_DB} || die "Could not create test database"

# public user role
PUBLICUSER=`node -e "console.log(require('${TESTENV}').db_pubuser || 'xxx')"`
PUBLICPASS=`node -e "console.log(require('${TESTENV}').db_pubuser_pass || 'xxx')"`
echo "DROP USER IF EXISTS ${PUBLICUSER};" | psql -v ON_ERROR_STOP=1 ${TEST_DB} || exit 1
echo "CREATE USER ${PUBLICUSER} WITH PASSWORD '${PUBLICPASS}';" | psql -v ON_ERROR_STOP=1 ${TEST_DB} || exit 1

# db owner role
TESTUSER=`node -e "console.log(require('${TESTENV}').db_user || '')"`
if test -z "$TESTUSER"; then
  echo "Missing db_user from ${TESTENV}" >&2
  exit 1
fi
TESTUSER=`echo ${TESTUSER} | sed "s/<%= user_id %>/${TESTUSERID}/"`
TESTPASS=`node -e "console.log(require('${TESTENV}').db_user_pass || '')"`
TESTPASS=`echo ${TESTPASS} | sed "s/<%= user_id %>/${TESTUSERID}/"`
echo "DROP USER IF EXISTS ${TESTUSER};" | psql -v ON_ERROR_STOP=1 ${TEST_DB} || exit 1
echo "CREATE USER ${TESTUSER} WITH PASSWORD '${TESTPASS}';" | psql -v ON_ERROR_STOP=1 ${TEST_DB} || exit 1


if test x"$OPT_COVERAGE" = xyes; then
  echo "Running tests with coverage"
  ./node_modules/.bin/istanbul cover ./node_modules/.bin/_mocha -- -u tdd -t 5000 $@ test/**/*.js
else
  echo "Running tests"
  ./node_modules/.bin/mocha -t 5000 -u tdd $@ test/**/*.js
fi
exit $?


lib/psql.js
'use strict';

var _ = require('underscore');
var debug = require('debug')('cartodb-psql');
var QueryWrapper = require('./query_wrapper');
var pg = require('pg');//.native; // disabled for now due to: https://github.com/brianc/node-postgres/issues/48

var stdTypeName = require('./oid-to-name');
// Holds a typeId->typeName mapping for each database ever connected to
var extTypeName = {};

pg.on('error', function () {
    // All errors will be handled synchronously or at client level
    // this avoid uncaughtException
});

// Workaround for https://github.com/Vizzuality/CartoDB-SQL-API/issues/100
var types = pg.types;
var arrayParser = pg.types.arrayParser;
var floatArrayParser = function(val) {
    if(!val) { return null; }
    var p = arrayParser.create(val, function(entry) {
        return parseFloat(entry);
    });
    return p.parse();
};
types.setTypeParser(20, parseFloat); // int8
types.setTypeParser(700, parseFloat); // float4
types.setTypeParser(701, parseFloat); // float8
types.setTypeParser(1700, parseFloat); // numeric
types.setTypeParser(1021, floatArrayParser); // _float4
types.setTypeParser(1022, floatArrayParser); // _float8
types.setTypeParser(1231, floatArrayParser); // _numeric
types.setTypeParser(1016, floatArrayParser); // _int8

/**
 * A simple postgres wrapper with logic about username and database to connect
 * - intended for use with pg_bouncer
 * - defaults to connecting with a "READ ONLY" user to given DB if not passed a specific user_id
 *
 * @param {Object} connectionParams Connection param options
 * - user: database username
 * - pass: database user password
 * - host: database host
 * - port: database port
 * - dbname: database name
 * @param {Object} poolParams
 * - size
 * - idleTimeout
 * - reapInterval
 * @param {Object} options
 * - {Number} maxRowSize The max number of bytes for a row, when exceeded the query will throw an error.
 * @returns PSQL
 */
var PSQL = function(connectionParams, poolParams, options) {
    options = options || {};
    options.maxRowSize = options.maxRowSize || undefined;

    var me = {
        POOL_DEFAULT_SIZE: 16,
        POOL_DEFAULT_IDLE_TIMEOUT: 3000,
        POOL_DEFAULT_REAP_INTERVAL: 1000
    };

    // default pool params by global settings or default value
    var globalSettings = global.settings || {};
    var _poolParams = {
        size: globalSettings.db_pool_size || me.POOL_DEFAULT_SIZE,
        idleTimeout: globalSettings.db_pool_idleTimeout || me.POOL_DEFAULT_IDLE_TIMEOUT,
        reapInterval: globalSettings.db_pool_reapInterval || me.POOL_DEFAULT_REAP_INTERVAL
    };

    // pool params will have precedence over global or default settings
    poolParams = poolParams || {};
    _poolParams = _.extend(_poolParams, poolParams);

    // Max database connections in the pool
    // Subsequent connections will block waiting for a free slot
    pg.defaults.poolSize = _poolParams.size;

    // Milliseconds of idle time before removing connection from pool
    pg.defaults.poolIdleTimeout = _poolParams.idleTimeout;

    // Frequency to check for idle clients within the pool, ms
    pg.defaults.reapIntervalMillis = _poolParams.reapInterval;

    // Max row size returned by PG stream
    pg.defaults.maxRowSize = options.maxRowSize || globalSettings.db_max_row_size;

    // keep alive configuration
    var keepAliveConfig;

    if (globalSettings.db_keep_alive) {
        keepAliveConfig = globalSettings.db_keep_alive.enabled;
    }

    var error_text = "Incorrect access parameters." +
        " If you are accessing via OAuth, please check your tokens are correct." +
        " For public users, please ensure your table is published.";
    if ( ! connectionParams || ( !_.isString(connectionParams.user) && !_.isString(connectionParams.dbname))) {
        throw new Error(error_text);
    }

    me.dbopts = connectionParams;

    me.poolParams = _poolParams;

    me.username = function(){
        return this.dbopts.user;
    };

    me.password = function(){
        return this.dbopts.pass;
    };

    me.database = function(){
        return this.dbopts.dbname;
    };

    me.dbhost = function(){
        return this.dbopts.host;
    };

    me.dbport = function(){
        return this.dbopts.port;
    };

    me.conString = "tcp://";
    if (me.username()) {
        me.conString += me.username();
    }
    me.conString += ":";
    if (me.password()) {
        me.conString += me.password();
    }
    me.conString += "@";
    if (me.dbhost()) {
        me.conString += me.dbhost();
    }
    if (me.dbport()) {
        me.conString += ":" + me.dbport();
    }
    me.conString += "/" + me.database();

    me.connectionObject = {
        host: me.dbhost(),
        port: me.dbport(),
        database: me.database(),
        user: me.username(),
        password: me.password(),
        keepAlive: keepAliveConfig,
        ssl: false
    };

    me.getConnectionConfig = function () {
        if (globalSettings.db_use_config_object) {
            return me.connectionObject;
        }
        return me.conString;
    };

    me.dbkey = function(){
        return this.database() + ":" + this.dbhost() + ":" + me.dbport();
    };

    me.ensureTypeCache = function(cb) {
        var db = this.dbkey();
        if (extTypeName[db]) {
            return cb();
        }
        this.dbConnect(this.getConnectionConfig(), function(err, client, done) {
            if (err) {
                return cb(err);
            }
            var types = ["'geometry'","'raster'","'geography'"]; // types of interest
            var typesSqlQuery = "SELECT oid, typname FROM pg_type where typname in (" + types.join(',') + ")";
            client.query(typesSqlQuery, function(err,res) {
                done();
                if (err) {
                    return cb(err);
                }
                var cache = {};
                res.rows.map(function(r) {
                    cache[r.oid] = r.typname;
                });
                extTypeName[db] = cache;
                cb();
            });
        });
    };

    // Return type name for a type identifier
    //
    // Possibly returns undefined, for unkonwn (uncached)
    //
    me.typeName = function(typeId) {
        return stdTypeName[typeId] ? stdTypeName[typeId] : extTypeName[this.dbkey()][typeId];
    };

    me.typeModInfo = function(typmod) {

        /* See https://github.com/postgis/postgis/blob/
           1277c750cb41807eefee4438a711746042d0cffe/liblwgeom/liblwgeom.h.in#L81-L98 */
        var wkbTypes = [
            "Unknown", "Point", "LineString", "Polygon",
            "MultiPoint", "MultiLineString", "MultiPolygon", "GeometryCollection",
            "CircularString", "CompoundCurve", "CurvePolygon", "MultiCurvePolygon",
            "PolyhedralSurface", "Triangle", "TIN"
        ];

        /* See https://github.com/postgis/postgis/blob/
           1277c750cb41807eefee4438a711746042d0cffe/liblwgeom/liblwgeom.h.in#L164-L172 */
        var srid   = ((typmod & 0x0FFFFF00) - (typmod & 0x10000000)) >> 8;
        var lwtype = (typmod & 0x000000FC)>>2;
        var has_z  = (typmod & 0x00000002)>>1;
        var has_m  = (typmod & 0x00000001);
        var ndims  = 2 + has_z + has_m;
        var wkbtype = wkbTypes[lwtype];

        return { srid: srid, ndims: ndims, wkbtype: wkbtype };
    };

    me.dbConnect = function(conConfig, cb) {
        pg.connect(conConfig, function(err, client, done) {
            if ( err ) {
                debug("PostgreSQL connection error: %s - connection config: %s", err, conConfig);
                err = new Error("cannot connect to the database");
                err.http_status = 500; // connection errors are internal
            }
            cb(err, client, done);
        });
    };

    me.connect = function(cb){
        var that = this;
        this.ensureTypeCache(function(err) {
            if (err) {
                return cb(err);
            }

            that.dbConnect(that.getConnectionConfig(), cb);
        });
    };

    me.eventedQuery = function(sql, params, callback){
        var that = this;
        if (typeof params === 'function') {
          callback = params;
          params = [];
        }
        params = params || [];

        this.connect(function(err, client, done) {
            if (err) {
                return callback(err);
            }

            var query = client.query(sql, params);

            function maxRowSizeListener(message) {
                pg.cancel(that.getConnectionConfig(), client, query);
                query.emit('error', new Error('Row too large, was ' + message.length + ' bytes'));
            }

            // forward notices to query
            var noticeListener = function() {
                query.emit('notice', arguments);
            };
            client.on('notice', noticeListener);

            var gotError = false;
            query.on('error', function(err) {
                client.removeListener('maxRowSize', maxRowSizeListener);
                client.removeListener('notice', noticeListener);
                gotError = true;
                done(err);
            });

            client.once('maxRowSize', maxRowSizeListener);

            // NOTE: for some obscure reason passing "done" directly
            //       as the listener works but can be slower
            //      (by x2 factor!)
            query.on('end', function() {
                client.removeListener('maxRowSize', maxRowSizeListener);
                client.removeListener('notice', noticeListener);
                if (!gotError) {
                    done();
                }
            });

            var queryCanceller = function() {
                pg.cancel(undefined, client, query);
            };

            return callback(null, query, queryCanceller);
        });
    };

    me.quoteIdentifier = function(str) {
        return pg.Client.prototype.escapeIdentifier(str);
    };

    me.escapeLiteral = function(str) {
        return pg.Client.prototype.escapeLiteral(str);
    };


    /**
     * Executes an array of queries.
     * Returns error if any query fails
     * Returns the result of the last query (if successful)
     */
    function queryArray(client, sql, params, callback) {
        client.query(sql[0], params, (err, result) => {
            sql.shift();
            if (err) {
                return callback(err);
            }
            if (sql.length === 0) {
                return callback(err, result);
            }

            return queryArray(client, sql, params, callback);
        });
    }

    /**
     *
     * @param {String} sql - Query
     * @param {Object} params - Query parameters
     * @param {Function} callback - (error, result)
     * @param {Boolean} readonly - If set, the query will be READ ONLY
     * @param {Integer} timeout - Milliseconds to timeout the query
     */
    me.query = function(sql, params, callback, readonly, timeout) {
        if (typeof params === 'function') {
          timeout = readonly;
          readonly = callback;
          callback = params;
          params = [];
        }

        this.connect(function(err, client, done) {
            if (err) {
                return callback(err);
            }
            if (!!readonly) {
                sql = `SET TRANSACTION READ ONLY; ${sql}`;
            }

            let sql_array = [ sql ];
            if (timeout !== undefined) {
                sql_array.unshift(`SET statement_timeout=${timeout};`);
            }

            queryArray(client, sql_array, params, (err, result) => {
                done(err);
                return callback(err, result);
            });
        });
    };

    me.end = function (callback) {
        callback = callback || function dummyCallback() {};

        var count = Object.keys(pg._pools).length;

        pg.once('end', function () {
            debug(count + ' pools were shutting down successfully');
            callback();
        });

        pg.end();
    };

    return me;
};

module.exports = PSQL;
module.exports.QueryWrapper = QueryWrapper;


lib/query_wrapper.js
'use strict';

var _    = require('underscore'),
    util = require('util'),
    SORT_ORDER_OPTIONS = {ASC: 1, DESC: 1},
    REGEX_SELECT = /^\s*SELECT\s/i,
    REGEX_INTO = /\sINTO\s+([^\s]+|"([^"]|"")*")\s*$/i;

function QueryWrapper(sql) {
    this.sqlQuery = sql.replace(/;\s*$/, '');
    this.sqlClauses = {
        orderBy: '',
        limit: ''
    };
}

/**
 * Only window select functions (NOTE: "values" will be broken, "with" will be broken)
 *
 * @param {number} limit
 * @param {number} offset
 * @returns {QueryWrapper}
 */
QueryWrapper.prototype.window = function (limit, offset) {
    if (!_.isNumber(limit) || !_.isNumber(offset)) {
        return this;
    }
    this.sqlClauses.limit = util.format(' LIMIT %d OFFSET %d', limit, offset);
    return this;
};

/**
 *
 * @param {string} column The name of the column to sort by
 * @param {string} sortOrder Whether it's ASC or DESC ordering
 * @returns {QueryWrapper}
 */
QueryWrapper.prototype.orderBy = function (column, sortOrder) {
    if (!_.isString(column) || _.isEmpty(column)) {
        return this;
    }
    this.sqlClauses.orderBy = util.format(' ORDER BY "%s"', column);

    if (!_.isUndefined(sortOrder)) {
        sortOrder = sortOrder.toUpperCase();
        if (SORT_ORDER_OPTIONS[sortOrder]) {
            this.sqlClauses.orderBy += util.format(' %s', sortOrder);
        }
    }
    return this;
};

/**
 * Builds an SQL query with extra clauses based on the builder calls.
 *
 * @returns {string} The SQL query with the extra clauses
 */
QueryWrapper.prototype.query = function () {
    if (_.isEmpty(this.sqlClauses.orderBy) && _.isEmpty(this.sqlClauses.limit)) {
        return this.sqlQuery;
    }
    // Strip comments
    this.sqlQuery = this.sqlQuery.replace(/--.*(\n|$)/g, ' ');

    var cte = '';

    if (this.sqlQuery.match(/^\s*WITH\s/i)) {

        var rem = this.sqlQuery, // analyzed portion of sql
            q, // quote char
            n = 0, // nested parens level
            s = 0, // 0:outQuote, 1:inQuote
            l;
        while (1) {
            l = rem.search(/[("')]/);
            // console.log("REM Is " + rem);
            if (l < 0) {
                // console.log("Malformed SQL");
                return this.sqlQuery;
            }
            var f = rem.charAt(l);
            // console.log("n:" + n + " s:" + s + " l:" + l + " charAt(l):" + f + " charAt(l+1):" + rem.charAt(l+1));
            if (s === 0) {
                if (f === '(') {
                    ++n;
                } else if (f === ')') {
                    if (!--n) { // end of CTE
                        cte += rem.substr(0, l + 1);
                        rem = rem.substr(l + 1);
                        //console.log("Out of cte, rem is " + rem);
                        if (rem.search(/^s*,/) < 0) {
                            break;
                        } else {
                            continue; // cte and rem already updated
                        }
                    }
                } else { // enter quoting
                    s = 1;
                    q = f;
                }
            } else if (f === q) {
                if (rem.charAt(l + 1) === f) {
                    ++l; // escaped
                } else {
                    s = 0; // exit quoting
                }
            }
            cte += rem.substr(0, l + 1);
            rem = rem.substr(l + 1);
        }
        /*
         console.log("cte: " + cte);
         console.log("rem: " + rem);
         */
        this.sqlQuery = rem; //sql.substr(l+1);
    }

    //console.log("SQL " + sql);
    //console.log(" does " + ( sql.match(REGEX_SELECT) ? '' : 'not ' ) + "match REGEX_SELECT " + REGEX_SELECT);
    //console.log(" does " + ( sql.match(REGEX_INTO) ? '' : 'not ' ) + "match REGEX_INTO " + REGEX_INTO);

    if (this.sqlQuery.match(REGEX_SELECT) && !this.sqlQuery.match(REGEX_INTO)) {
        return util.format(
            '%sSELECT * FROM (%s) AS cdbq_1%s%s',
            cte, this.sqlQuery, this.sqlClauses.orderBy, this.sqlClauses.limit
        );
    }

    return cte + this.sqlQuery;
};

module.exports = QueryWrapper;


lib/oid-to-name.js
module.exports = {
    16: 'bool',
    17: 'bytea',
    18: 'char',
    19: 'name',
    20: 'int8',
    21: 'int2',
    22: 'int2vector',
    23: 'int4',
    24: 'regproc',
    25: 'text',
    26: 'oid',
    27: 'tid',
    28: 'xid',
    29: 'cid',
    30: 'oidvector',
    71: 'pg_type',
    75: 'pg_attribute',
    81: 'pg_proc',
    83: 'pg_class',
    114: 'JSON',
    142: 'xml',
    143: '_xml',
    199: '_json',
    194: 'pg_node_tree',
    32: 'pg_ddl_command',
    210: 'smgr',
    600: 'point',
    601: 'lseg',
    602: 'path',
    603: 'box',
    604: 'polygon',
    628: 'line',
    629: '_line',
    700: 'float4',
    701: 'float8',
    702: 'abstime',
    703: 'reltime',
    704: 'tinterval',
    705: 'unknown',
    718: 'circle',
    719: '_circle',
    790: 'money',
    791: '_money',
    829: 'macaddr',
    869: 'inet',
    650: 'cidr',
    1000: '_bool',
    1001: '_bytea',
    1002: '_char',
    1003: '_name',
    1005: '_int2',
    1006: '_int2vector',
    1007: '_int4',
    1008: '_regproc',
    1009: '_text',
    1028: '_oid',
    1010: '_tid',
    1011: '_xid',
    1012: '_cid',
    1013: '_oidvector',
    1014: '_bpchar',
    1015: '_varchar',
    1016: '_int8',
    1017: '_point',
    1018: '_lseg',
    1019: '_path',
    1020: '_box',
    1021: '_float4',
    1022: '_float8',
    1023: '_abstime',
    1024: '_reltime',
    1025: '_tinterval',
    1027: '_polygon',
    1033: 'aclitem',
    1034: '_aclitem',
    1040: '_macaddr',
    1041: '_inet',
    1263: '_cstring',
    1042: 'bpchar',
    1043: 'varchar',
    1082: 'date',
    1083: 'time',
    1114: 'timestamp',
    1115: '_timestamp',
    1182: '_date',
    1183: '_time',
    1184: 'timestamptz',
    1185: '_timestamptz',
    1186: 'interval',
    1187: '_interval',
    1231: '_numeric',
    1266: 'timetz',
    1270: '_timetz',
    1560: 'bit',
    1561: '_bit',
    1562: 'varbit',
    1563: '_varbit',
    1700: 'numeric',
    1790: 'refcursor',
    2201: '_refcursor',
    2202: 'regprocedure',
    2203: 'regoper',
    2204: 'regoperator',
    2205: 'regclass',
    2206: 'regtype',
    4096: 'regrole',
    4089: 'regnamespace',
    2207: '_regprocedure',
    2208: '_regoper',
    2209: '_regoperator',
    2210: '_regclass',
    2211: '_regtype',
    4097: '_regrole',
    4090: '_regnamespace',
    2950: 'uuid',
    2951: '_uuid',
    3220: 'pg_lsn',
    3221: '_pg_lsn',
    3614: 'tsvector',
    3642: 'gtsvector',
    3615: 'tsquery',
    3734: 'regconfig',
    3769: 'regdictionary',
    3643: '_tsvector',
    3644: '_gtsvector',
    3645: '_tsquery',
    3735: '_regconfig',
    3770: '_regdictionary',
    3802: 'jsonb',
    3807: '_jsonb',
    2970: 'txid_snapshot',
    2949: '_txid_snapshot',
    3904: 'int4range',
    3905: '_int4range',
    3906: 'numrange',
    3907: '_numrange',
    3908: 'tsrange',
    3909: '_tsrange',
    3910: 'tstzrange',
    3911: '_tstzrange',
    3912: 'daterange',
    3913: '_daterange',
    3926: 'int8range',
    3927: '_int8range',
    2249: 'record',
    2287: '_record',
    2275: 'cstring',
    2276: 'any',
    2277: 'anyarray',
    2278: 'void',
    2279: 'trigger',
    3838: 'event_trigger',
    2280: 'language_handler',
    2281: 'internal',
    2282: 'opaque',
    2283: 'anyelement',
    2776: 'anynonarray',
    3500: 'anyenum',
    3115: 'fdw_handler',
    3310: 'tsm_handler',
    3831: 'anyrange',
};